만국박람회 [STEP 1] Maxhyunm, Whales

WhalesJin opened the pull request:
안녕하세요 Steven(@stevenkim18)🦦!!
2주동안 Steven께 만국박람회 프로젝트 리뷰를 받게된 Max🤖, Whales🐋 입니다!!
UI를 직접 만들어보는 것도, JSON 데이터를 받아오는 것도 다 처음이라 이번 프로젝트가 많이 걱정되면서도 그만큼 설렙니다!!!
아직 부족한 부분이 많지만 보시고 조언주신다면, 더 공부해서 더 성장하도록 해보겠습니다!!

잘 부탁드립니다 🙇🏻‍♀️



# STEP 1
## 🤹🏻 고민했던 점
### 🔹 Codable / Decodable / Encodable
`JSON` 인코딩/디코딩을 위해서는 `Codable`, `Decodable`, `Encodable` 프로토콜 중 선택하여 채택해야 하는 것으로 이해하고 있습니다. 데이터를 받기와 보내기가 모두 이루어지는 경우라면 `Codable`을 활용해야 하지만, 만국박람회의 경우 저장된 `JSON 데이터`를 불러오기만 하는 것으로 생각되어 `Decodable`을 선택하였습니다.

### 🔹 let / var
처음에는 `ExpositionUniverselle`과 `Items` 구조체의 프로퍼티를 모두 `var`로 생성하려고 했습니다. 하지만 한 번 불러온 데이터를 수정할 일이 있을까?라는 생각이 들었습니다. 결과적으로 만국박람회 데이터는 특별히 전처리를 거치며 수정할 내용이 없고, 기존의 데이터를 보여주기만 하면 충분한 것으로 판단하여 `let`으로 결정하게 되었습니다.

### 🔹 CodingKey
데이터셋에 포함된 `JSON 파일`은 스위프트와 달리 `스네이크 케이스`로 작성되어 있었습니다. 때문에 `카멜 케이스` 형식으로 변환하기 위하여 중첩 `CodingKey` `Enum`을 구현했습니다

### 🔹 UIImage
`Items` 구조체에서 `imageName`을 활용하여 `UIImage`를 불러오는 연산 프로퍼티를 추가하려고 했습니다. 하지만 이번 요구사항은 모델 타입을 구현하는 것이고, `UIImage`를 불러오는 로직은 테이블 뷰에서 수행하는 게 맞는 것 같다는 생각이 들어 추후로 미뤄두었습니다.

### 🔹 NumberFormatter  
숫자 타입으로 되어있는 방문객 수를 천단위 구분기호가 들어간 `String` 타입으로 리턴하기 위해 `NumberFormatter`를 사용했습니다. 처음에는 `NumberFormatter`를 위한 파일을 구분하려 했는데, 다음과 같은 이유에서 `ExpositionUniverselle` 구조체 내부에서 연산 프로퍼티로 처리했습니다.
- 프로젝트 내부에서 달리 포매터가 사용되는 것이 없음
- 만약 박람회가 여러 개로 확장이 되더라도 디코딩시 동일한 타입을 사용하게 된다면 해당 프로퍼티로 충분히 활용 가능

WhalesJin committed:
feat: Items 구조체 생성

- JSON 이름 형식을 바꾸기 위한 CodingKeys 열거형 생성

maxhyunm committed:
feat: ExpositionUniverselle 구조체 생성 및 Items 리팩토링

Items에서 Codable 대신 Decodable 프로토콜 채택
Items 프로퍼티를 var에서 let으로 변경

WhalesJin committed:
feat: Items Test 파일 생성

- JSON의 Decoding이 잘 되는지 테스트

maxhyunm committed:
feat: ExpositionUniverselle Test 생성

- JSON의 Decoding이 잘 되는지 테스트

WhalesJin committed:
feat: ExpositionUniverselle 구조체에 formattedVisitors 프로퍼티 생성

- 방문자 수에 천단위 구분기호를 넣기 위해서 NumberFormatter 사용

maxhyunm committed:
feat: formattedVisitors Test 생성

- 구분기호 여부 확인용 테스트 생성
- 테스트 실패로 누락된 NumberFormatter의 numberStyle 설정 발견하여 추가

stevenkim18 commented on Expo1900/Expo1900/Model/ExpositionUniverselle.swift:
```diff
@@ -0,0 +1,27 @@
+
+import Foundation
+
+struct ExpositionUniverselle: Decodable {
```
해당 모델 이름에 대해서 크게 문제 되지는 않지만 조금 더 고민해볼까요?

이 타입이 박람회 정보를 담고 있는 모델 혹은 정보다 라는 것을 이름에 더 구체적으로 알 수 있으면 좋을 것 같은데.. 
더 좋은 네이밍이 있을까요?

WhalesJin replied:
fccbdfe7ed3daadd489f1ce9698e0fba6c8c3677
주신 리뷰를 참고하여 박람회 정보의 뜻을 담아 `ExpositionData`로 네이밍 수정하였습니다!!😄

stevenkim18 replied:
음.. 사실 Data도 좀 범용적으로 사용되어서 살짝 애매한 네이밍 같네요 ㅎㅎ

저는 보통 json으로 부터 받아오는 원천 데이트는 뒤에 entity를 붙이는 편입니다.
이번에 바꾸실 필요는 없는데 
entity에 대해서 한 번 살펴보세요!

stevenkim18 commented on Expo1900/Expo1900/Model/ExpositionUniverselle.swift:
```diff
@@ -0,0 +1,27 @@
+
+struct ExpositionUniverselle: Decodable {
+    let title: String
+    let visitors: UInt
```
`Uint`로 선언하신 이유가 있나요?

WhalesJin replied:
방문자 수는 0 이상의 정수로 나올 것이라 생각하여
`UInt` 타입으로 선언하였습니다!!

stevenkim18 commented on Expo1900/Expo1900/Model/ExpositionUniverselle.swift:
```diff
@@ -0,0 +1,27 @@
+    let duration: String
+    let description: String
+    
+    var formattedVisitors: String {
+        let numberFormatter = NumberFormatter()
+        numberFormatter.numberStyle = .decimal
+        
+        guard let result = numberFormatter.string(for: visitors) else {
+            return String(visitors)
+        }
+        
+        return result
+    }
```
지금은 물론 이곳에 밖에 쓰이지 않지만
`Formatter` 같은 종류의 전역으로 쓰이는 경우는 따로 코드를 분리하는 것이 좋습니다.

어떻게 분리하면 좋을까요?
(저는 개인적으로 extension으로 분리하는 것을 좋아합니다)

maxhyunm replied:
5a305dfc4409a119e845c327879ea50eb6b9d399 피드백 주신 덕분에 UInt extension으로 변경 완료하였습니다!

stevenkim18 commented on Expo1900/Expo1900/Model/Items.swift:
```diff
@@ -0,0 +1,20 @@
+//  Created by Max, Whales on 2023/06/27.
+//
+
+struct Items: Decodable {
```
타입의 이름만 봐서는 무엇을 뜻하는지 모르겠어요.

어떤 Item을 말하는 건가요?? 

네이밍이 더 구체적이고 명확하면 좋을 것 같아요!

WhalesJin replied:
fccbdfe7ed3daadd489f1ce9698e0fba6c8c3677
박람회의 출품작을 뜻하는 `Item`이고 이 구조체도 출품작의 정보를 담고 있다고 생각하여
`ExpositionItemData`로 네이밍 수정하였습니다!! 😃

stevenkim18 replied:
여기도 Data를 사용한 네이밍은 한 번 더 고민해보세요 ㅎㅎ

stevenkim18 commented on Expo1900/Expo1900/Model/Items.swift:
```diff
@@ -0,0 +1,20 @@
+    let shortDescription: String
+    let description: String
+    
+    enum CodingKeys: String, CodingKey {
+        case name
+        case imageName = "image_name"
+        case shortDescription = "short_desc"
+        case description = "desc"
+    }
```
`CodingKey`를 아주 잘 사용하셨네요 ㅎㅎ
좋습니다! 👍

stevenkim18 commented on Expo1900/Expo1900ModelTests/Expo1900ExpositionUniverselleTests.swift:
```diff
@@ -0,0 +1,48 @@
+        sut = nil
+    }
+
+    func test_개최지_이름이_프랑스_파리이다() {
+        // given
+        let expectation = "프랑스 파리"
+        
+        // when
+        let location = sut.location
+        
+        // then
+        XCTAssertEqual(expectation, location)
+    }
+    
+    func test_방문자_수에_구분기호가_들어간다() {
+        // given
+        let expectation = "48,130,300"
+        
+        // when
+        let formattedVisitors = sut.formattedVisitors
+        
+        // then
+        XCTAssertEqual(expectation, formattedVisitors)
+    }
+}
```
해당 테스트는 무엇을 테스트를 하는 것 일까요?
테스트를 하는 주체가 무엇인가요?

위 테스트의 경우 전시회 이름이나 방문자 수가 바뀌면 어떻게 테스트 할 수 있을까요??

maxhyunm replied:
5a305dfc4409a119e845c327879ea50eb6b9d399 공유해주신 내용 참고하여 테스트 내용 수정하였습니다! ExpositionData, ExpositionItemData의 경우 mock 객체를 활용하였고 UInt는 NumberFormatter 위주의 테스트만 추가하였습니다.

stevenkim18 reviewed (commented):
안녕하세요!🙋‍♂️ 맥스, 웨일!! 저도 잘 부탁드립니다! 🙇‍♂️

처음 들어오셨을 때 PR 관련 필독사항에도 적혀있듯이 
코드리뷰는 저와 의견을 나누는 자리입니다!!

코드에는 정답은 없습니다!! 제가 틀리다면 근거를 가지고 지적해주세요!!

질문을 받는 저도 성장하는 시간입니다!! **많이 질문해주세요!!** ㅎㅎ

저는 **커뮤니케이션**을 매우 중요하게 생각합니다. 실제로 현업에서 매우 중요하다고 생각합니다.

- PR을 보낼 때 어떻게 하면 최대한 **2분이서 짠 코드를 효과적으로 보여줄 수 있을지**를 고민해주세요!!
    - 최대한 글을 **가독성** 있게 작성해주시고 글로 부족하다면 저에게 음성 쳇을 요청해주세요!
- 질문을 할 때 이 **질문을 한 배경**과 **고민한 점**을도 같이 남겨주시면 제가 더 많은 이야기를 해드릴 수 있습니다.
- 코딩 뿐만 아니라 글, 말 등도 많이 훈련하셔서 여기서 많은 것들을 얻어가셨으면 합니다! ㅎㅎ

---

## 고민한 점에 대한 답변

> 🔹 Codable / Decodable / Encodable
> 
> 
> `JSON` 인코딩/디코딩을 위해서는 `Codable`, `Decodable`, `Encodable` 프로토콜 중 선택하여 채택해야 하는 것으로 이해하고 있습니다. 데이터를 받기와 보내기가 모두 이루어지는 경우라면 `Codable`을 활용해야 하지만, 만국박람회의 경우 저장된 `JSON 데이터`를 불러오기만 하는 것으로 생각되어 `Decodable`을 선택하였습니다.
> 

사실 저는 여기서 `Codable` 를 사용해도 큰 무리는 없을 것 같다는 생각이 들어요.

다만 정확하게`Decodable` 적어야 하는 이유는 **나중에 이 코드를 보고 분석하기 위함입니다.**

현업에 있다보면 기존에 작성 되어 있는 코드를 분석해야 합니다. 

그럴 때 저 사소한 차이가 큰 시간을 벌어주게 됩니다. 

`Codable` 을 채택하게 되면 코드를 읽는 입장에서는 “아 이 모델은 encoding 하는 코드도 있겠구나”라고 생각을 할 수도 있곘죠? 인코딩하는 로직을 찾다가 없다는 것을 알기 까지 꽤 오랜 시간이 걸릴 수도 있어요 ㅋㅋ

그래서 작성을 할 때 최대한 정확하게 작성하고 필요하지 않는 코드들은 잘 삭제를 해두는 습관을 들이는 것이 좋습니다 ㅎㅎ

> 🔹 let / var
> 
> 
> 처음에는 `ExpositionUniverselle`과 `Items` 구조체의 프로퍼티를 모두 `var`로 생성하려고 했습니다. 하지만 한 번 불러온 데이터를 수정할 일이 있을까?라는 생각이 들었습니다. 결과적으로 만국박람회 데이터는 특별히 전처리를 거치며 수정할 내용이 없고, 기존의 데이터를 보여주기만 하면 충분한 것으로 판단하여 `let`으로 결정하게 되었습니다.
> 

저는 보통 let으로 선언하고 빌드를 돌려서 error가 나면 var로 바꾸는 편입니다. ㅋㅋ

하지만 더 좋은 것은 미리 객체와 모델들을 잘 설계 하게 되면 let var를 미리 정할 수 있습니다.

> 🔹 UIImage
> 
> 
> `Items` 구조체에서 `imageName`을 활용하여 `UIImage`를 불러오는 연산 프로퍼티를 추가하려고 했습니다. 하지만 이번 요구사항은 모델 타입을 구현하는 것이고, `UIImage`를 불러오는 로직은 테이블 뷰에서 수행하는 게 맞는 것 같다는 생각이 들어 추후로 미뤄두었습니다.
> 
> 🔹 NumberFormatter
> 
> 숫자 타입으로 되어있는 방문객 수를 천단위 구분기호가 들어간 `String` 타입으로 리턴하기 위해 `NumberFormatter`를 사용했습니다. 처음에는 `NumberFormatter`를 위한 파일을 구분하려 했는데, 다음과 같은 이유에서 `ExpositionUniverselle` 구조체 내부에서 연산 프로퍼티로 처리했습니다.
> 
> - 프로젝트 내부에서 달리 포매터가 사용되는 것이 없음
> - 만약 박람회가 여러 개로 확장이 되더라도 디코딩시 동일한 타입을 사용하게 된다면 해당 프로퍼티로 충분히 활용 가능

위 글에 대해서는 기술적으로 코멘트 할 것은 없습니다.

다만 읽는 사람이 조금 더 빠르게 이해 할 수 있게 적어주시면 더 좋을 것 같아요!

물론 코드를 보면 바로 이해가 가겠지만, PR 특성상 PR 내용이 먼저 보이고 코드는 그 뒤에 보이기 때문에 관련 코드를 첨부해주셨으면 글을 이해하는데 더 빨랐을 것 같습니다.

아니면 코드를 볼 수 있는 commit 링크도 괜찮겠네요.😃

https://github.com/WhalesJin/ios-exposition-universelle/commit/540425bede0c33a4304bf9389ec47fb911381bf4

## Git Commit에 대한 코멘트

커밋을 보다 보니 조금 더 분리 했으면 하는 커밋이 있어서 코멘트 남겨요 ㅎㅎ

https://github.com/WhalesJin/ios-exposition-universelle/commit/540425bede0c33a4304bf9389ec47fb911381bf4

위 커밋에 내용 “feat: Items 구조체 생성” 인데 

json과 asset 파일들이 해당 커밋에 함께 들어 있네요.

프로젝트에 필요한 파일 커밋은 따로 분리해주셔도 좋을 듯 합니다.

나중에 해당 커밋을 검색해서 찾아볼 때 Items 구조체 관련 코드를 볼 것 같은데..

이미지와 json 파일들이 엄청 많으면 “이게 뭐야?” 할 것 같습니다😂

WhalesJin committed:
refactor: 타입명  수정

- ExpositionUniverselle을 ExpositionData로 수정
- Items를 ExpositionItemData로 수정

maxhyunm committed:
refactor: NumberFormatter 로직 위치 변경, 테스트 수정

- NumberFormatter 로직을 UInt 타입에 formatToDecimal 메서드로 구현
- ExpositionData, ExpositionItemData의 테스트를 mock 객체를 활용하도록 수정
- UInt formatToDecimal 테스트 추가

stevenkim18 commented:
이번 스텝은 간단해서 바로 다음 스텝으로 넘어가겠습니다. 고생하셨습니다💪🏻
